File: src/player.js
'use strict';

const fs = require('fs');
const path = require('path');
const { spawn } = require('child_process');
const { send } = require('./ipc');
const { ipcPath, stateDir, mpvLogFile, logFile, debugEnabled } = require('./paths');

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

function log(message) {
  if (!debugEnabled()) return;
  try {
    fs.mkdirSync(stateDir(), { recursive: true });
    const time = new Date().toISOString().slice(11, 23);
    fs.appendFileSync(logFile(), `${time} pid=${process.pid} ${message}\n`);
  } catch {
  }
}

async function waitAlive() {
  for (let i = 0; i < 30; i += 1) {
    if (await alive()) {
      log(`  waitAlive: socket up after ${i * 100}ms`);
      return true;
    }
    await sleep(100);
  }
  log('  waitAlive: TIMED OUT after 3000ms');
  return false;
}

function mpvBin() {
  return process.env.VIBECODE_MPV_BIN || 'mpv';
}

async function alive() {
  const reply = await send(ipcPath(), { command: ['get_property', 'mpv-version'] });
  return !!(reply && reply.error === 'success');
}

function buildArgs(source, volume) {
  const args = [
    '--no-video',
    '--no-terminal',
    '--really-quiet',
    '--idle=yes',
    '--keep-open=yes',
    '--loop-playlist=inf',
    '--network-timeout=15',
    '--stream-lavf-o=reconnect=1,reconnect_at_eof=1,reconnect_streamed=1,reconnect_on_network_error=1,reconnect_delay_max=2',
    '--cache=yes',
    '--demuxer-max-bytes=1MiB',
    '--demuxer-readahead-secs=20',
    '--audio-stream-silence=yes',
    '--audio-wait-open=1',
    `--volume=${volume}`,
    '--pause',
    `--input-ipc-server=${ipcPath()}`,
  ];
  if (debugEnabled()) {
    args.push(`--log-file=${mpvLogFile()}`, '--msg-level=all=status');
  }
  if (process.env.VIBECODE_MPV_ARGS) {
    for (const flag of process.env.VIBECODE_MPV_ARGS.split(/\s+/)) {
      if (flag) args.push(flag);
    }
  }
  args.push(source);
  return args;
}

async function start(source, volume) {
  const lockFile = path.join(stateDir(), 'starting');
  try {
    if (Date.now() - fs.statSync(lockFile).mtimeMs < 10000) {
      return waitAlive();
    }
  } catch {
  }
  try {
    fs.mkdirSync(stateDir(), { recursive: true });
    fs.writeFileSync(lockFile, String(process.pid));
  } catch {
  }

  if (process.platform !== 'win32') {
    try {
      fs.unlinkSync(ipcPath());
    } catch {
    }
  }
  try {
    log(`  start: spawning mpv (${mpvBin()}) source=${source}`);
    const child = spawn(mpvBin(), buildArgs(source, volume), {
      detached: true,
      stdio: 'ignore',
      windowsHide: true,
    });
    child.on('error', (e) => log(`  start: spawn error ${e && e.code}`));
    child.unref();
    return await waitAlive();
  } catch {
    return false;
  } finally {
    try {
      fs.unlinkSync(lockFile);
    } catch {
    }
  }
}

module.exports = { alive, start, mpvBin };

File: src/player.rs
use crate::ipc;
use crate::log::log;
use crate::paths;
use std::env;
use std::fs;
use std::process::{Command, Stdio};
use std::thread::sleep;
use std::time::{Duration, SystemTime};

fn mpv_bin() -> String {
    env::var("VIBECODE_MPV_BIN").unwrap_or_else(|_| "mpv".into())
}

pub fn alive() -> bool {
    ipc::send(&serde_json::json!({ "command": ["get_property", "mpv-version"] }))
        .and_then(|r| {
            r.get("error")
                .and_then(|e| e.as_str())
                .map(|s| s == "success")
        })
        .unwrap_or(false)
}

fn wait_alive() -> bool {
    for i in 0..30 {
        if alive() {
            log(&format!("  wait_alive: socket up after {}ms", i * 100));
            return true;
        }
        sleep(Duration::from_millis(100));
    }
    log("  wait_alive: TIMED OUT after 3000ms");
    false
}

fn build_args(source: &str, volume: i64) -> Vec<String> {
    let mut args: Vec<String> = [
        "--no-video",
        "--no-terminal",
        "--really-quiet",
        "--idle=yes",
        "--keep-open=yes",
        "--loop-playlist=inf",
        "--network-timeout=15",
        "--stream-lavf-o=reconnect=1,reconnect_at_eof=1,reconnect_streamed=1,reconnect_on_network_error=1,reconnect_delay_max=2",
        "--cache=yes",
        "--demuxer-max-bytes=1MiB",
        "--demuxer-readahead-secs=20",
        "--audio-stream-silence=yes",
        "--audio-wait-open=1",
    ]
    .iter()
    .map(|s| s.to_string())
    .collect();
    args.push(format!("--volume={volume}"));
    args.push("--pause".into());
    args.push(format!("--input-ipc-server={}", paths::ipc_path()));
    if paths::debug_enabled() {
        args.push(format!("--log-file={}", paths::mpv_log_file().display()));
        args.push("--msg-level=all=status".into());
    }
    if let Ok(extra) = env::var("VIBECODE_MPV_ARGS") {
        for flag in extra.split_whitespace() {
            args.push(flag.to_string());
        }
    }
    args.push(source.to_string());
    args
}

#[cfg(windows)]
pub fn spawn_detached(cmd: &mut Command) -> std::io::Result<()> {
    use std::os::windows::process::CommandExt;
    cmd.creation_flags(0x0800_0008); // DETACHED_PROCESS | CREATE_NO_WINDOW
    cmd.spawn().map(|_| ())
}

#[cfg(unix)]
pub fn spawn_detached(cmd: &mut Command) -> std::io::Result<()> {
    use std::os::unix::process::CommandExt;
    cmd.process_group(0);
    cmd.spawn().map(|_| ())
}

pub fn start(source: &str, volume: i64) -> bool {
    let lock = paths::state_dir().join("starting");
    if let Ok(meta) = fs::metadata(&lock) {
        if let Ok(modified) = meta.modified() {
            if SystemTime::now()
                .duration_since(modified)
                .is_ok_and(|d| d.as_secs() < 10)
            {
                return wait_alive();
            }
        }
    }
    let _ = fs::create_dir_all(paths::state_dir());
    let _ = fs::write(&lock, std::process::id().to_string());

    #[cfg(unix)]
    let _ = fs::remove_file(paths::ipc_path());

    log(&format!(
        "  start: spawning mpv ({}) source={}",
        mpv_bin(),
        source
    ));
    let mut cmd = Command::new(mpv_bin());
    cmd.args(build_args(source, volume))
        .stdin(Stdio::null())
        .stdout(Stdio::null())
        .stderr(Stdio::null());
    let ok = spawn_detached(&mut cmd).is_ok();
    let result = if ok { wait_alive() } else { false };
    let _ = fs::remove_file(&lock);
    result
}
